src/main/java/org/apache/commons/compress/archivers/zip/ZipFile.java

```diff
@@ -368,9 +369,14 @@ public static void closeQuietly(final ZipFile zipFile) {
     private final List<ZipArchiveEntry> entries = new LinkedList<>();
 
     /**
-     * Maps String to list of ZipArchiveEntrys, name -> actual entries.
+     * Maps a string to the first entry by that name.
      */
-    private final Map<String, LinkedList<ZipArchiveEntry>> nameMap = new HashMap<>(HASH_SIZE);
+    private final Map<String, ZipArchiveEntry> nameMap = new HashMap<>(HASH_SIZE);
+
+    /**
+     * If multiple entries have the same name, maps the name to entries by that name.
+     */
+    private Map<String, List<ZipArchiveEntry>> duplicateNameMap;
```

Comment:
Given that this is extremely unlikely, would it make sense to keep the original Map and change the value to Object and let it hold either ZipArchiveEntry or the List?  It would eventually complicate the lookup but that should be a single place only.

Reply:
I  REALLY don't like that. When you look at the pack200 code, which is written at the pre-generics Java version, and some of its lists, you can't tell if some of the lists are buggy, sloppy, or trying to be clever. That or the lack of generics let's you get away with not having fully thought through the hierarchy of types you put in those lists.

Reply:
That's true, such optimizations typically come at cost of readability and safety.  However, keeping the amount of code touching the data structures limited may help - we'll need only three methods - `first-by-name`, `all-by-name` and `add-entry` .  And document the field to not touch except via those three methods :-)

Just a suggestion if it comes to true optimization, given this PR is purely about improving time and space performance.